src/Access/ContextAccess.cpp

```diff
@@ -299,6 +299,7 @@ void ContextAccess::setUser(const UserPtr & user_) const
         current_roles = user->granted_roles.findGranted(params.current_roles);
         current_roles_with_admin_option = user->granted_roles.findGrantedWithAdminOption(params.current_roles);
     }
+    current_roles.insert(current_roles.end(), params.external_roles.begin(), params.external_roles.end());
```

Comment:
It doesn't feel right to add roles to `current_roles` without adding them to `user->default_roles`. I think we can just put external roles instead of current roles, and modify `User` in `LDAPAccessStorage` to contain the default roles we want.